materialize-sql/templating.go

```diff
@@ -64,56 +63,85 @@ type MergeBound struct {
 // MergeBoundsBuilder tracks and generates a MergeBound for each of a binding's
 // key fields.
 type MergeBoundsBuilder struct {
-	table     Table
-	literaler func(any) string
+	keyColumns []Column
+	literaler  func(any) string
 
-	lower tuple.Tuple
-	upper tuple.Tuple
+	lower []any
+	upper []any
 }
 
-func NewMergeBoundsBuilder(table Table, literaler func(any) string) *MergeBoundsBuilder {
+func NewMergeBoundsBuilder(keyColumns []Column, literaler func(any) string) *MergeBoundsBuilder {
 	return &MergeBoundsBuilder{
-		table:     table,
-		literaler: literaler,
+		keyColumns: keyColumns,
+		literaler:  literaler,
 	}
 }
 
-// NextStore updates the observed minimum and maximum key for a transaction. It
-// relies on the fact that Stores are sent to materializations in ascending key
-// order, so the first observed Store will have the minimum key and the final
-// observed store will have the maximum key.
-func (b *MergeBoundsBuilder) NextStore(key tuple.Tuple) {
-	if len(key) != len(b.table.Keys) {
-		panic(fmt.Sprintf("application error: %d key fields vs. %d key columns for merge query bounds", len(key), len(b.table.Keys)))
+// NextKey updates the observed minimum and maximum key for a transaction.
+func (b *MergeBoundsBuilder) NextKey(key []any) {
+	if len(key) != len(b.keyColumns) {
+		panic(fmt.Sprintf("application error: %d key fields vs. %d key columns for merge query bounds", len(key), len(b.keyColumns)))
 	}
 
 	if b.lower == nil {
-		b.lower = key
+		b.lower = append([]any(nil), key...)
```

Comment:
is this `append` just copying the data? I don't really understand it

Reply:
I added a comment here to explain it a little better. Let me know if that makes sense!

I had initially tried this with `b.lower = key` and `b.upper = key`, but then they both point to the same slice and that doesn't work for tracking the lower and upper bound separately. So there's this initial copy operation to initialize the values for each transaction, which then will be updated as needed by the code just below here.

The existing tests cover this since `b.lower` is `nil` when initializing the `MergeBoundsBuilder` struct.

Reply:
I thought we would have updated snapshots for this scenario that's why I asked about test case